pkg/objectstore/bucket.go

```diff
@@ -210,9 +213,34 @@ func (p *s3Provider) DeleteBucket(ctx context.Context, bucketName string) error
 	return location.RemoveContainer(bucketName)
 }
 
-// returns the region for a particular bucket
-func (p *s3Provider) getRegionForBucket(ctx context.Context, bucketName string) (string, error) {
-	return GetS3BucketRegion(ctx, bucketName, p.config.Region)
+// GetRegionForBucketreturns the region for a particular bucket. It does not
+// the region set in the provider config is used as a hint, but may differ than
+// the actual region of the bucket. If the bucket does not have a region, then
+// the return value will be "",
+func (p *s3Provider) GetRegionForBucket(ctx context.Context, bucketName string) (string, error) {
+	cfg, r, err := awsConfig(ctx, p.config, *p.secret.Aws)
+	if err != nil {
+		return "", err
+	}
+	s, err := session.NewSession(cfg)
+	if err != nil {
+		return "", errors.Wrapf(err, "failed to create session, region = %s", r)
+	}
+	cli := s3.New(s)
+	if cli == nil {
+		return "", errors.New("failed to create s3 client")
+	}
+	gbli := &s3.GetBucketLocationInput{
+		Bucket: aws.String(bucketName),
+	}
+	gblo, err := cli.GetBucketLocation(gbli)
```

Comment:
I worry about using `GetBucketLocation` over `GetBucketRegion` (from the `s3manager`) package.

`GetBucketLocation` is restricted to the owner of the bucket. This means that if a bucket is shared between two accounts, only the owning account will be able to make this call. Other accounts won't be able to determine the region for the bucket.